Add AWS CDK docs

KNiepok opened the pull request:
# Description of the change

My take on AWS CDK docs that explains how AWS CDK can be used inside a CloudFormation stack. 


## Checklist

Please make sure that the proposed change checks all the boxes below before requesting a review:

- [x] I have reviewed the [guidelines for contributing](https://github.com/spacelift-io/user-documentation/blob/main/CONTRIBUTING.md) to this repository.
- [x] The preview looks fine.
- [x] The tests pass.
- [x] The commit history is clean and meaningful.
- [x] The pull request is opened against the `main` branch.
- [x] The pull request is no longer marked as a draft.
- [x] You agree to license your contribution under the [MIT license](https://github.com/spacelift-io/user-documentation/blob/main/LICENSE) to Spacelift.

If the proposed change is ready to be merged, please request a review from `@spacelift-io/solutions-engineering`. Someone will review and merge the pull request.

Thank you for your contribution! 🙇

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+# Integrating with AWS Cloud Development Kit (CDK)
+
+To use [AWS Cloud Development Kit](https://docs.aws.amazon.com/cdk/v2/guide/home.html){: rel="nofollow"} in an AWS CloudFormation Stack you'll need to do two things: create a Docker image with CDK included and invoke CDK in `before_plan` hooks.
```
```suggestion
To use [AWS Cloud Development Kit](https://docs.aws.amazon.com/cdk/v2/guide/home.html){: rel="nofollow"} in an AWS CloudFormation stack you'll need to do two things: create a Docker image with AWS CDK and invoke it in `before_plan` hooks.
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+## Preparing the image
+
+Since our base image doesn't support CDK, you will have to build your image that includes `cdk` as well as any tooling needed to run whatever language your infrastructure declaration is written in.
```
```suggestion
Since our base image doesn't support AWS CDK, you will have to build your image that includes it as well as any tooling needed to run whatever language your infrastructure declaration is written in.
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+# Install packages
+RUN apk update && apk add --update --no-cache \
+    python3 \
```
`python3` is already in the base image.

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+# Install AWSCLI
+RUN pip install --upgrade pip && \
+    pip install --upgrade awscli
```
`awscli` is already in the base image.

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+## Adding `before_plan` hooks
+
+For the CDK code to be properly interpreted by Spacelift, you have to customize the default stack workflow by [enriching them with hooks](https://docs.spacelift.io/concepts/configuration/runtime-configuration/index.html#before_-and-after_-hooks)
```
```suggestion
For the AWS CDK code to be properly interpreted by Spacelift, you have to customize the default stack workflow by [enriching them with hooks](https://docs.spacelift.io/concepts/configuration/runtime-configuration/index.html#before_-and-after_-hooks)
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+## Adding `before_plan` hooks
+
+For the CDK code to be properly interpreted by Spacelift, you have to customize the default stack workflow by [enriching them with hooks](https://docs.spacelift.io/concepts/configuration/runtime-configuration/index.html#before_-and-after_-hooks)
+If your CDK project is not yet bootstrapped, you should add:
```
```suggestion
If your AWS CDK project is not yet bootstrapped, you should add:
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
````diff
@@ -0,0 +1,67 @@
+For the CDK code to be properly interpreted by Spacelift, you have to customize the default stack workflow by [enriching them with hooks](https://docs.spacelift.io/concepts/configuration/runtime-configuration/index.html#before_-and-after_-hooks)
+If your CDK project is not yet bootstrapped, you should add:
+```
+cdk bootstrap
````
Maybe this should be inlined as `cdk bootstrap` instead.

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
````diff
@@ -0,0 +1,67 @@
+as your `before_plan` hook.
+To create a CloudFormation template that can be interpreted by Spacelift, you will have to add a
+```
+cdk synth
````
Maybe this should be inlined as `cdk synth` instead.

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
````diff
@@ -0,0 +1,67 @@
+```
+cdk synth
+```
+hook to the `before_plan` stage.
````
Also, we may want to just list the commands as bullet points since they are to be added to the same hook.

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+The default CloudFormation integration in Spacelift uses a single CloudFormation template.
+That means that CDK definitions that generate multiple templates will only have a single template picked up for further processing.
+To mitigate this, consider unifying the CDK definition to generate a single template file.
```
```suggestion
To mitigate this, consider unifying the AWS CDK definition to generate a single template file.
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+### Deploying Lambdas
+
+Our integration doesn't use `cdk deploy`, but rather uses template definitions created by `cdk synth`.
+`cdk deploy` deploys lambda assets to s3 which are used to deploy lambdas by CloudFormation.
```
```suggestion
`cdk deploy` deploys Lambda assets to S3 which are used to deploy Lambdas by CloudFormation.
```

jmfontaine commented on docs/vendors/cloudformation/integrating-with-cdk.md:
```diff
@@ -0,0 +1,67 @@
+
+Our integration doesn't use `cdk deploy`, but rather uses template definitions created by `cdk synth`.
+`cdk deploy` deploys lambda assets to s3 which are used to deploy lambdas by CloudFormation.
+Our process won't upload assets, so deploying lambdas via a Spacelift Stack configured to handle CDK will result in errors.
```
```suggestion
Our process won't upload assets, so deploying Lambdas via a Spacelift stack configured to handle AWS CDK will result in errors.
```

jmfontaine commented on docs/vendors/cloudformation/reference.md:
```diff
@@ -8,7 +8,7 @@ description: Details about all available AWS CloudFormation-specific configurati
 - **Region** - AWS region in which to create and execute the AWS CloudFormation Stack.
 - **Stack Name** - The name for the CloudFormation Stack controlled by this Spacelift Stack.
-- **Entry Template File** - The path to the JSON or YAML file describing your root CloudFormation Stack. If you're generating CloudFormation code using a tool like AWS Serverless Application Model (SAM), point this to the file containing the generated template.
+- **Entry Template File** - The path to the JSON or YAML file describing your root CloudFormation Stack. If you're generating CloudFormation code using a tool like AWS Serverless Application Model (SAM) or AWS Cloud Development Kit (CDK), point this to the file containing the generated template.
```
```suggestion
- **Entry Template File** - The path to the JSON or YAML file describing your root CloudFormation Stack. If you're generating CloudFormation code using a tool like AWS Serverless Application Model (SAM) or AWS Cloud Development Kit (AWS CDK), point this to the file containing the generated template.
```

jmfontaine reviewed (commented):
I suggested a few minor changes but otherwise, it looks good to me. 👍

KNiepok commented:
@jmfontaine thanks for insighful review, I fixed the content according to the comments 🙇

jmfontaine commented:
@KNiepok I see 3 recommended changes that have not been implemented but I see no comment to push back on them. Am I missing something?

KNiepok commented:
@jmfontaine oh, I think this is because I used my editor rather than applying changes via Github and I've also squashed and rebased the commits. Next time I will use the Github mechanism to make it more visible. 
But I've checked the content again and all of the suggestions you gave are already applied.

jmfontaine commented:
@KNiepok GitHub's UI is correct. I checked the branch out locally and it is missing some if the fixes I recommended.

![CleanShot 2023-02-15 at 11 51 29](https://user-images.githubusercontent.com/174728/219097675-e4d3a816-122d-4060-990b-a5551d5f789f.png)

Maybe you committed those changes but did not push them?

jmfontaine reviewed (approved):
Looks good to me!
